pkg/search/postgres/store_cache.go

```diff
@@ -408,6 +369,37 @@ func (c *cachedStore[T, PT]) GetIDsByQuery(ctx context.Context, query *v1.Query)
 	return c.underlyingStore.GetIDsByQuery(ctx, query)
 }
 
+func (c *cachedStore[T, PT]) getByQueryFn(ctx context.Context, query *v1.Query, fn func(obj PT) error) error {
+	if query == nil || query.EqualVT(search.EmptyQuery()) {
+		c.cacheLock.RLock()
+		defer c.cacheLock.RUnlock()
+		return c.walkCacheNoLock(ctx, fn)
+	}
+
+	identifiers, err := c.underlyingStore.GetIDsByQuery(ctx, query)
+	// Fallback to the underlying store on error.
+	if err != nil {
+		log.Errorf("Failed to get identifiers by query, falling back to get results by query: %v", err)
+		return c.underlyingStore.GetByQueryFn(ctx, query, fn)
+	}
+	c.cacheLock.RLock()
+	defer c.cacheLock.RUnlock()
+	for _, id := range identifiers {
+		obj, found := c.cache[id]
+		if !found {
+			log.Warnf("Object %q not found in store cache", id)
+			continue
+		}
+		if !c.isReadAllowed(ctx, obj) {
+			continue
+		}
+		if err := fn(obj); err != nil {
```

Comment:
Would it make sense to work on a clone of the cached object in order to shield it from being altered by the passed function ? This should apply to walkCacheNoLock as well.
This may have the issue that clusters and namespaces lookups are then slower and could impact the performance of scope computation at the beginning of each service call.

Reply:
The cluster lookup (as well as the namespace one) relies on store `Walk`, which does `walkCacheNoLock` under lock.
Here, the function can be applied directly on a clone of the cached object.
In the case of the walkCacheNoLock at the start of the function, I think it would be good to wrap the function in a function that clones the cached object and passes the clone to the parameter function.

Reply:
I added a test to check if function operates on a clone